### DISCUSSION TITLE
Gradient masks also kind of work

### DISCUSSION THREAD

**Discussion by SMUsamaShah** [zero01101/openOutpaint#28] [Ideas]
I tried to fade out the mask going towards original image.

This is the initial image I am using. 
![landscape](https://user-images.githubusercontent.com/385283/202965343-234b026c-ef2b-45b8-ab99-b3a630a35c11.png)

I test this by outpainting the right half.
It's a liner gradient from `black` to `white`. Simply replacing lines from https://github.com/zero01101/openOutpaint/blob/d1a131a9b844b46a31dc4a434f88bd48256ab156/js/index.js#L651 to https://github.com/zero01101/openOutpaint/blob/d1a131a9b844b46a31dc4a434f88bd48256ab156/js/index.js#L670

with

```javascript
let grd = overMaskCanvasCtx.createLinearGradient(0, 0, 255, 0);
grd.addColorStop(0, "black");
grd.addColorStop(1, "white");
overMaskCanvasCtx.fillStyle = grd;
overMaskCanvasCtx.fillRect(0, 0, 254, 512);
```

This is what the mask looks like
![image](https://user-images.githubusercontent.com/385283/202966526-5e0ec76f-607e-4919-ae25-0571ee221f87.png)

Problem is that it only affects half of the original image while making the outpainted portion black.
![image](https://user-images.githubusercontent.com/385283/202966390-a86789df-d488-4bf2-88ec-248932c3c1d2.png)

Idea behind trying this was that any outpaint should also effect original image a little bit depending on whats in the outpainted half of the image.

**Comment by SMUsamaShah** [comment]
My bad, it should have been `overMaskCanvasCtx.fillRect(0, 0, 512, 512);`

Generated 2 images
![image](https://user-images.githubusercontent.com/385283/202969020-233b4229-0708-4975-be25-2b9e03e5e0ac.png)
second image is kind of a failure though
![image](https://user-images.githubusercontent.com/385283/202969219-eae416e7-008b-4637-afd5-cb9c0902a7af.png)

**Reply by zero01101** [reply]
yeah; for any future curious types, gotta keep that alpha channel opaque for it to actually do anything :)

**Comment by zero01101** [comment]
y'know i tried gradients when i was being stupid with overmasking and they always gave me a really blatantly visible, well, gradient seam but i honestly don't think i tried it _wide_ enough maybe? certainly something to potentially play with more...

and yeah, i sometimes get bonus "windowpaned" images in my outpaints regardless of overmask/blur, legitimately NO clue where to even start with that one lol

**Reply by SMUsamaShah** [reply]
> i sometimes get bonus "windowpaned" images in my outpaints regardless of overmask/blur, legitimately NO clue where to even start with that one lol

I believe the fix for that is randomness :) Same thing you are already doing when masking the edges with random arcs. Reason it should work is that because gradient is still a straight linear line, SD sees that as an opportunity to draw a new image from somewhere in that spectrum. If its random, SD won't see any straight line and hence no new image.

Will have to try again by adding random noise in the gradient. 
(i have removed the auto repo+models after these experiments again to cure my addiction with SD)

**Reply by SMUsamaShah** [reply]
Using noisy mask does not work well. It makes the image blurry, loosing finer details.

**Reply by seijihariki** [reply]
What about noisy gradient masks?

**Comment by SMUsamaShah** [comment]
Skewing the gradient is giving much better results e.g. turning it kind logarithmic instead of linear.

```
grd.addColorStop(0, "rgba(0,0,0,255)");
grd.addColorStop(0.25, "rgba(128,128,128,255)");
grd.addColorStop(1, "rgba(255,255,255,255)");
```

![image](https://user-images.githubusercontent.com/385283/202971171-7a7a5481-60ab-45c4-b7cc-db9ce4aea2ce.png)

EDIT: Overall, the good thing is, gradient mask is not producing any visible seams.

EDIT: Realised that making huge changes in large portion of original image will not be desirable in most cases.

EDIT: This kind of gradient is effecting the image the way I wanted to, slight changes as a whole, more near the edges 

```
grd.addColorStop(0, "rgba(0,0,0,255)");
grd.addColorStop(0.1, "rgba(120,120,120,255)");
grd.addColorStop(0.8, "rgba(120,120,120,255)");
grd.addColorStop(1, "rgba(255,255,255,255)");
```

**Reply by zero01101** [reply]
i don't know what your prompt was for this but it's very [arrival](https://en.wikipedia.org/wiki/Arrival_(film)) and i _love_ it

**Reply by SMUsamaShah** [reply]
I think it was something like "alien spaceship with glowing lights, award winning". Was trying to see the effect of adding lights on original image.
 Love that film, will probably be watching again soon.

**Comment by seijihariki** [comment]
I have been testing this, and from my tests at least, what gradient does is not really a gradient at all... Seems that it simply does not consider anything blacker than something around #808080 as a mask, and so keeps what is there. Anything over this number seems to generate fully, without regard for the brightness.
I'm checking in the webui code because maybe I'm missing something, and I believe there should be a way to support gradients, but for now it unfortunately seems in automatic's it doesn't really do anything.

I suggest you test with fully gray masks instead of a gradient, that should give more of an idea of what different gray tones do...

**Reply by SMUsamaShah** [reply]
I did.

And that's how I ended up with the last skewed gradient. What happens if you use gray is that what looks like an untouched image on gray mask actually has slightly different color. This leaves a very visible seam.

For example in above image using gray mask makes right half of the mountain a slightly different color. 

I believe that below a certain threshold the changes in images are not structural and so seamless that you won't notice unless you have something to compare.

So, gradient below that value does work, just not as we are expecting it to.

**Reply by SMUsamaShah** [reply]
Gradients are already supported by automatic1111. That's how I got them working. Or may be I misunderstood what you are saying here about support.

**Reply by seijihariki** [reply]
Well, if you say so, it shouldn't be so difficult to implement some gradient editing tools.

If that's accurate, some of the features (img2img border masking in particular) could probably greatly benefit from this.

I will add a separate branch for gradient editing support and we can see what to do from there.
